lib/internal/crypto/x509.js

```diff
@@ -202,6 +203,15 @@ class X509Certificate {
     return value;
   }
 
+  get isValid() {
+    let value = this[kInternalState].get('isValid');
+    if (value === undefined) {
+      value = this[kHandle].isValid();
+      this[kInternalState].set('isValid', value);
+    }
+    return value;
+  }
```

Comment:
Could this be implemented in terms of `.validFrom`/`.validTo`, rather than adding a new internal accessor? If the format of `.validFrom`/`.validTo` isn't helpful for that, would it make more sense to expose those times in a more standard format (e.g. JS `Date` objects) instead?

Reply:
I guess so, but like you have said we would probably require some sort of parsing mechanism in JS for the format of these fields. Most of these accessors are simply returning what OpenSSL gives you straight away, so I was hesitant on changing `.validFrom` / `.validTo` to something else from the current ASN.1 time formatted string.(I would imagine that for users familiar with OpenSSL, they would expect these values to be in the same format, with perhaps it enabling some other convenient interactions with external systems using the said format. However I'm not familiar with such use cases on this in general, so I would love to see more opinions on this)
If that seems like a more fitting change for Node to expose `Date` objects instead of raw strings from OpenSSL, then I suppose that could be a better solution instead of adding a new field.

With that said, the implementation of `X509_cmp_timeframe` itself also isn't trivial as far as I can tell, so I'm thinking that this is the most straightforward way for this to be implemented without providing such conversions of `.validFrom` and `.validTo` to javascript `Date` or similar objects.

Reply:
> If that seems like a more fitting change for Node to expose Date objects instead of raw strings from OpenSSL, then I suppose that could be a better solution instead of adding a new field.

I would be tempted to lean in this direction if possible (and explore what's required to make correct date validation practical - I'm not sure what mysteries might lie within `X509_cmp_timeframe`). I'm fairly cautious about `isValid` as a specific name, since that implies a broader range of certificate validation than just start/end date checks (which all this code is doing). There are many contexts in which people may assume a 'valid' certificate is a much more specific thing.

Either directly exposing `Date` fields or using different terminology (temporally valid? `isCurrent`? `hasExpired`+`wasBeenIssuedInThePast`?) would clear that up nicely.

Oh, and also there's conceptually an inevitable race condition here, since this depends on the current time, which users might make different kinds of assumptions about (most wrongly/problematically: does this mean it was valid _at the time I received it?_ No). Either exposing date fields (so users think about the comparison themselves) or exposing an API that always takes a date to check (`isCurrent(date)`) would remove any possible confusion there.

Reply:
Aha, I suppose something like `isExpiredAt(date)` would be more appropriate here. Thank you for the explanation.

I will try to look into parsing the time string into a javascript object first and see where I end up with.